### ISSUE TITLE
Add support for different deep learning frameworks in Fabrik

### ISSUE THREAD

**Issue by utsavgarg** [Cloud-CV/GSoC-Ideas#14]
**Description:** [Fabrik](https://github.com/Cloud-CV/Fabrik) is an online collaborative web application for building and visualizing neural networks in the browser, aimed at lowering the barrier to entry to getting started with deep learning. We wish to create a model-agnostic platform which works with most popular deep learning frameworks and provides a seamless experience when trying to visualize existing models, create new ones, or to export them to the framework of your choice.

Over the last couple of years, we have built a version of this platform which provides a drag-and-drop interface for creating neural networks, extensively supports two popular deep learning frameworks (Caffe and Keras), and has experimental support for Tensorflow, another widely-popular framework.

Our goal moving forward is to complete support for our existing frameworks and extend support to new and upcoming frameworks like PyTorch. Initially, we would like to accomplish the following tasks:
* Complete support for Tensorflow (see [#293](https://github.com/Cloud-CV/Fabrik/issues/293))
* Add support for PyTorch (see [#294](https://github.com/Cloud-CV/Fabrik/issues/294))
* Add an exhaustive documentation for new and existing backends
* Write robust unit tests

**Deliverables:** The issues mentioned above should be implemented by the end of the GSoC period. Moreover, we expect students to also have implemented improvements that they have proposed.

**Extended Goals:**
* Setup docker containers for easy setup on local machines (see [#130](https://github.com/Cloud-CV/Fabrik/issues/130))
* Extend login functionality to allow users to privately share their networks with others
* Investigate the addition of [ONNX](https://github.com/onnx/onnx) support
* Improve the visualization algorithm to make it more generalized.

**Mentors:** [Utsav Garg](https://utsavgarg.github.io/)

**Co-Mentor(s):** [Deshraj Yadav](https://deshraj.github.io/)

**Skills:**
* *Essential:* React JS, Python, Django, Unix.
* *Preferred:* Some experience with Tensorflow, PyTorch, Keras

**Skill Level:** Medium

**Getting started:** Take a look at our [issues](https://github.com/Cloud-CV/Fabrik/issues) on Github, the ones marked as `starter-project` are good places to start. Feel free to reach out to us on our [Gitter](https://github.com/Cloud-CV/Fabrik) channel if you have questions.

**Important links:**
- Fabrik Website: http://fabrik.cloudcv.org
- Fabrik Github: https://github.com/Cloud-CV/Fabrik
- Fabrik Docs: http://fabrik.readthedocs.io/en/latest/
- GSoC Proposal Template: https://github.com/Cloud-CV/GSoC-Ideas/wiki/GSOC-2018-Proposal-Template
- Gitter Channel: https://gitter.im/Cloud-CV/
- Mailing list: https://groups.google.com/forum/#!forum/cloudcv

**Comment by kapilkd13** [comment]
Hi @utsavgarg @deshraj  I am kapil. I found this project really interesting and relevant and would love to work on it as part of GSOC 2018. Last year I participated with Open Bioinformatics Foundation and worked on CWL- a workflow management tool. I have decent experience with TF and Pytorch and have been developing in Python since last couple of years. 
It looks like this issue https://github.com/Cloud-CV/Fabrik/issues/119 is being worked on. Could you guys suggest some other starter issue I can look upon, meanwhile I will setup Fabrik on my system. Thanks :smiley:

**Comment by migom6** [comment]
Hi, 
I am Madhurjya Pegu, currently studying in Indian Institute Of Information Technology, Allahabad, India.
I am a full stack developer and I am familiar with the React ecosystem (redux/mobx/mst, react-native or any of the Elm based arch). 
Since your project is based on web technologies, I am quite good at hacking things way out. 
I have used both Pytorch and tensorflow and I am quite familiar with their docs.
Thank you for your time.
**My github:- https://github.com/migom6**

**Comment by vasu-dev** [comment]
Hi @utsavgarg  @deshraj ,
I'm Vasudev Singh Yadav , a junior year undergrad at Delhi Technological University. I found this project interesting and would like to contribute to the same. I have been practising ML for around 1.5 years and is well versed in various deep learning frameworks and have good knowledge of Django.
Could you please suggest how to start.
Thanks for your time.

**Comment by 5anthosh** [comment]
Hi @utsavgarg @deshraj ,
I am Santhosh Kumar, currently pursuing bachelor degree in computer science in India. I am Deep learning enthusiast and I love to design & build DNN .I would love to contribute this project . I taught myself machine learning and am looking  forward to learn more. I also developed some web applications in Django and also familiar with other web technologies. I am also familiar with tensorflow 
thanks

**Comment by fenilsuchak** [comment]
Woah, this sounds interesting. I want like to contribute to this . Python , TF , Keras I work almost everyday. Django ocassionaly.
@utsavgarg

**Comment by chiral-carbon** [comment]
Hey @utsavgarg! 
I am a machine learning and deep learning enthusiast and novice, currently learning CNN and its applications. I am well-versed in Python and have worked on handwriting recognition and data classifiers, and loved this idea of making AI algos generally available and usable. I have worked with TensorFlow a little bit, but I'll pick it up quickly for this project. Would love to get the opportunity to work on it full-time! I have not really contributed much to open source projects but I see this as a great opportunity to dive into it.  
Thanks!

**Comment by tonysy** [comment]
Hi, @utsavgarg  
I am a graduate student major in deep learning and computer vision. Cloud-CV is a great platform, which could give huge help to scientist and student in this area. I am working  on image recognition topic, include classification, detection and segmentation. 
I am familiar with **Faster-RCNN** and **Mask-RCNN**, and I also have experience on image segmentation, by using **Deeplab** model. I have worked as a deep learning algorithm engineer in [TuSimple, Inc](http://tusimple.ai/), a startup company focus on **Autonomous Truck** , for a half year. I have used Tensorflow, PyTorch, MXNet, already for more than one year. I would like to join this project and work other friends, to make Cloud-CV more powerful. 
My website is: [syzhang.me](http://www.syzhang.me)
Thanks a lot.

**Comment by miqianmimi** [comment]
Hi, @utsavgarg
I am a senior student major in Data Science in Fudan University. Now i am doing my graduate project in HKUST, I am familiar with Machine Learning ,Computer Vision, i have post one paper on ACM MM as the third author with my professor YanweiFu. https://www.semanticscholar.org/paper/Learning-to-Generate-and-Edit-Hairstyles-Yin-Fu/b72045dbbdc35bc777186025190a99acf202a1e6
For more information about the project i have done , you can see from my website or viewing my application.:
www.miqianmimi.com
www.miqianmimi.wordpress.com
I would like to join Fabric and work with other friends, to make Cloud-CV more powerful.

Thanks a lot.

**Comment by utsavgarg** [comment]
Hi @kapilkd13 @migom6 @vasu-dev @santhoshkumarr1096 @Fenil3510 @chiral-carbon @tonysy @miqianmimi, thanks for your interest in CloudCV and particularly the Fabrik project. 

You could start by setting up the development environment and submitting fixes for [issues](https://github.com/Cloud-CV/Fabrik/issues) marked with `starter-project`. An in-depth knowledge of ML/DL/CV is not a requirement for contributing to this project, but some familiarity with the frameworks and common practices would be helpful to think about the importance of any feature. The project primarily uses React & Django and you should be at least familiar with their basics. Please follow https://github.com/Cloud-CV/GSoC-Ideas/wiki/GSOC-2018-Proposal-Template as a rough template for drafting your proposals.

Looking forward to your contributions and proposals.

**Comment by shujathkhan** [comment]
Hi @utsavgarg,
I'm Shujath, currently pursuing my final year at SRM University, I'm a machine learning and deep learning enthusiast. I'm also a contributor to open source projects.  I'm well-versed with React and Django. Would love to work on Fabrik and contribute to CloudCV as well.
You can check out my profile,
https://Shujathlive.github.io/backup-portfolio

Thank you!

**Comment by husnainfareed** [comment]
Hi, @utsavgarg @deshraj I am Hussnain. I found this project really interesting and relevant and would love to work on it as part of GSoC 2018. I have some recent experience with TF, Keras and Pytorch and have been developing in Python since last years. Would love to work on Fabrik and contribute to CloudCV as well.
Thanks 😃

**Comment by LuckyPigeon** [comment]
Hello @utsavgarg @deshraj I am a sophomore student who major in Computer Science in National Chi Nan University, I'm familiar with python, and take some course for AI, ML, DL using MXNET, and self learning using tensorflow and PyTorch for NLP and using OpenCV for CV, the laboratory of mine is web related, so I have some experience with sql using mysql and mssql, framework using drupal and web design using nodeJS, ReactJS, HTML5, CSS3. This is the first time that I'm participate in GSoC and would like to contribute to Fabrik for GSoC project, and contribute to CloudCV if I can as well, thanks!

**Comment by llucifer97** [comment]
Hi, @utsavgarg  @deshraj  I am second year student at Birla Institute of technology , I am good knowledge of framework like tensorflow and keras,and some experience in pytorch.I would like to solve this issue ,Please guide how to get started .

**Comment by RishabhJain2018** [comment]
Closing this issue since GSOC 2018 is over.
